src/com/google/enterprise/adaptor/database/UniqueKey.java

```diff
@@ -107,28 +111,38 @@
       tmpTypes.put(name, type);
     }
     names = Collections.unmodifiableList(tmpNames);
-    types = Collections.unmodifiableMap(tmpTypes);
+    Map<String, ColumnType> caseInsensitiveTypes = new TreeMap<>();
```

Comment:
What about using TreeMap<>(Strings.CASE_INSENSITIVE_ORDER) here, as verifyColumnNames does, then just caseInsensitiveTypes.putAll(tmpTypes), and if the sizes differ, types = tmpTypes, otherwise caseInsensitiveTypes. Then we can drop the caseSensitive field, and put splitIntoNameList back to static (as preferred when called from a constructor).